### ISSUE TITLE
Dtype strict mode

### ISSUE THREAD

**Issue by sinhrks** [pydata/pandas-design#19]
pandas ``dtype`` coercion is useful in most cases, but there are situations to prohibit it to avoid unexpected values being included. The behavior should be switchable like:

```
s = pd.Series([1, 2, 3])
s[1] = 1.2
s
# 0    1
# 1    1
# 2    3
# dtype: int64

s[1] = 'x'
s
# 0    1
# 1    x
# 2    3
# dtype: object

s = pd.Series([1, 2, 3], cast=False)
s[1] = 'x'
TypeError...
```

**Comment by MaximilianR** [comment]
Strict immutability is another potential option

**Comment by shoyer** [comment]
I agree with @MaximilianR, though I'm not sure if he was refers to immutable dtypes or immutable values altogether!

Dtype coercion is an feature that I would gladly drop altogether.

**Comment by MaximilianR** [comment]
> immutable dtypes or immutable values 

Immutable dtypes is the lower hanging fruit! But now you mention immutable values...

**Comment by wesm** [comment]
The implicit casting behavior that exists now (really just to permit nulls to be inserted into dtypes that don't support them) makes me super uncomfortable. I guess I would need to understand more use cases as adding the flexibility would be quite a bit of work.
